tests/ffn_model_tests.cpp

```diff
@@ -0,0 +1,82 @@
+/**
+ * @file model_tests.cpp
+ * @author Kartik Dutt
+ *
+ * Tests for various functionalities and performance of models.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#define BOOST_TEST_DYN_LINK
+#include <utils/utils.hpp>
+#include <ensmallen.hpp>
+#include <dataloader/dataloader.hpp>
+#include <models/darknet/darknet.hpp>
+#include <boost/test/unit_test.hpp>
+
+// Use namespaces for convenience.
+using namespace boost::unit_test;
+
+BOOST_AUTO_TEST_SUITE(FFNModelsTests);
+
+template<
+    typename OptimizerType,
+    typename OutputLayerType = mlpack::ann::NegativeLogLikelihood<>,
+    typename InitializationRuleType = mlpack::ann::RandomInitialization,
+    class MetricType = mlpack::metric::SquaredEuclideanDistance,
+    typename InputType = arma::mat,
+    typename OutputType = arma::mat
+>
+void CheckFFNClassificationWeights(mlpack::ann::FFN<OutputLayerType,
+    InitializationRuleType>& model, const std::string& datasetName,
+    const double threshold, const bool takeMean,
+    OptimizerType& optimizer)
+{
+  DataLoader<InputType, OutputType> dataloader(datasetName, true);
+
+  // Train the model. Note: Callbacks such as progress bar and loss aren't
+  // used in testing. Training the model for few epochs ensures that a
+  // user can use the pretrained model on any other dataset.
+  model.Train(dataloader.TrainFeatures(), dataloader.TrainLabels(), optimizer);
+
+  // Verify viability of model on validation datset.
+  OutputType predictions;
+  model.Predict(dataloader.ValidFeatures(), predictions);
+
+  // Since this checks weights for classification problem, we need to convert
+  // predictions into labels.
+  arma::Row<size_t> predLabels(predictions.n_cols);
+  for (arma::uword i = 0; i < predictions.n_cols; ++i)
+  {
+    predLabels(i) = predictions.col(i).index_max() + 1;
+  }
+
+  double error = MetricType::Evaluate(predLabels, dataloader.ValidLabels());
+
+  if (takeMean)
+  {
+    error = error / predictions.n_elem;
+  }
+
+  BOOST_REQUIRE_LE(error, threshold);
+}
+
+/**
+ * Simple test for Darknet model.
+ */
+BOOST_AUTO_TEST_CASE(DarknetModelTest)
+{
+  mlpack::ann::DarkNet<> darknetModel(3, 56, 56, 10);
+
+  // Create an optimizer object for tests.
+  ens::SGD<ens::AdamUpdate> optimizer(1e-4, 16, 50,
+      1e-8, true, ens::AdamUpdate(1e-8, 0.9, 0.999));
+
+  // Check whether FFN model performs well.
+  // CheckFFNClassificationWeights<ens::SGD<ens::AdamUpdate>>(
+  //    darknetModel1.GetModel(), "mnist", 1e-1, true, optimizer);
```

Comment:
Would it be a good to change `mnist` to `imagenet` now? and are there additional tests you want to add for DarkNet models?

Reply:
Sure, I have updated the tests but they aren't being run now. I have opened a PR with correct preprocessing required to convert mlpack loaded image equivalent to PyTorch's ToTensor(). We can use that to preprocess the data and use the saved model for prediction.

Reply:
Can you add that to this PR only and run the tests to make sure everything is working as intended?

Reply:
Yeah, already on it.

Reply:
I have added a simple tests as well as the imagenet preprocessor in this PR. I haven't added tests for weights because of mlpack/mlpack#2561. I think we can merge this PR since the model is working and any changes would be internal implementation in mlpack. To see if the model is working correctly you can run the following commands : 
```
git clone https://github.com/kartikdutt18/mlpack-PyTorch-Weight-Translator.git
cd mlpack-PyTorch-Weight-Translator
./run.sh
```
With this you will see the model weights being created and transferred into the Darknet model and being used for prediction. You can see that the final accuracy of `0.7236842105` (divided by 100). Kindly let me know if you face any issues. If this makes sense, I think we can merge this PR and then later add a PR with tests for weights once mlpack/mlpack#2561 is resolved.

Reply:
alright I will do a review. In the meantime can you work on finishing the YOLO PR ?

Reply:
Sure. I'll work on it.